grpc_adapter.go

```diff
@@ -0,0 +1,668 @@
+		grpcHeaderMessage: true,
+	}
+	for _, key := range strings.Split(header.Get(headerTrailer), ",") {
+		key = http.CanonicalHeaderKey(key)
+		isTrailer[key] = true
+	}
+
+	trailer := make(http.Header)
+	for key, vals := range header {
+		if strings.HasPrefix(key, http.TrailerPrefix) {
+			// Must remove trailer prefix before canonicalizing.
+			key = http.CanonicalHeaderKey(key[len(http.TrailerPrefix):])
+			trailer[key] = vals
+		} else if key := http.CanonicalHeaderKey(key); isTrailer[key] {
+			trailer[key] = vals
+		}
+	}
+
+	protobuf := w.config.Codec
+	trailerErr := grpcErrorFromTrailer(w.config.BufferPool, protobuf, trailer)
+	if trailerErr != nil {
+		if w.isStreaming() {
+			trailerErr.meta = trailer
+		}
+		return trailer, trailerErr
+	}
+
+	// Remove all protocol trailer keys.
+	for key := range trailer {
+		if isProtocolHeader(key) {
+			delete(trailer, key)
+		}
+	}
+	return trailer, nil
+}
+
+func (w *connectResponseWriter) finalizeUnary() error {
+	trailer, err := w.getTrailer()
+	if err != nil {
+		return err
+	}
+
+	// Encode trailers as header
+	responseHeader := w.ResponseWriter.Header()
+	for key, vals := range trailer {
+		if isProtocolHeader(key) {
+			continue
+		}
+		key = connectUnaryTrailerPrefix + key
+		responseHeader[key] = vals
+	}
+
+	_ = w.body.Next(5) // Discard envelope head.
+	contentLength := strconv.Itoa(w.body.Len())
+	setHeaderCanonical(responseHeader, headerContentLength, contentLength)
+
+	// Write buffered unary body to stream.
+	if _, err := w.body.WriteTo(w.ResponseWriter); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (w *connectResponseWriter) finalizeStream() error {
+	trailer, err := w.getTrailer()
+	if err != nil {
+		return err
+	}
+
+	// Encode as connect end message.
+	end := &connectEndStreamMessage{
+		Trailer: trailer,
+	}
+	data, err := json.Marshal(end)
+	if err != nil {
+		return err
+	}
+
+	head := [5]byte{connectFlagEnvelopeEndStream, 0, 0, 0, 0}
+	binary.BigEndian.PutUint32(head[1:5], uint32(len(data)))
+	if _, err := w.ResponseWriter.Write(head[:]); err != nil {
+		return err
+	}
+	if _, err := w.ResponseWriter.Write(data); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (w *connectResponseWriter) flushWithTrailers() {
+	if w.isStreaming() {
+		if err := w.finalizeStream(); err != nil {
+			responseHeader := w.ResponseWriter.Header()
+			setHeaderCanonical(responseHeader, headerContentType, w.typ)
+			_ = w.config.ErrorWriter.writeConnectStreaming(w.ResponseWriter, err)
+		}
+	} else {
+		if err := w.finalizeUnary(); err != nil {
+			responseHeader := w.ResponseWriter.Header()
+			delHeaderCanonical(responseHeader, headerContentLength)
+			delHeaderCanonical(responseHeader, connectUnaryHeaderCompression)
+			setHeaderCanonical(responseHeader, headerContentType, connectUnaryContentTypeJSON)
+			_ = w.config.ErrorWriter.writeConnectUnary(w.ResponseWriter, err)
+		}
+	}
+	flushResponseWriter(w.ResponseWriter)
+}
+
+func (w *connectResponseWriter) Flush() {
```

Comment:
To support Go 1.20's `ResponseController`, this type (and all impls of `ResponseWriter`) should also expose an `Unwrap` method:
```go
func (w *connectResponseWriter) Unwrap() http.ResponseWriter {
    return w.ResponseWriter
}
```
Ideally, we'd actually _check_ whether the underlying writer is a `http.Flusher` first, and then return a writer impl _without_ that method if it doesn't. As is, we just silently ignore this case in `flushResponseWriter`, but ideally the caller should be the one to decide what to do if the writer does not support flush.

In this case, [gRPC **requires** a writer that supports flush](https://github.com/grpc/grpc-go/blob/master/internal/transport/handler_server.go#L72-L76), so we should do the same: we should check for this at the top (in `NewGRPCAdapter`), instead of silently ignoring it when called.

Reply:
Added `ResponseController` interface!

For ConnectUnary Flushing isn't required and we want to avoid flushing the writer to capture trailers and re-encode as headers. I've added flusher checking to the other methods.